{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding genes within significant Fst windows\n",
    "Here we are using the Fst values we got from pairwise Fst analysis on GATK variants from the populations of Yemen, Ethiopia and India. The pairwise Fst values were calculated using Weir and Cockerham method (1984)\n",
    "As the first step we will find the windows with significantly high Fst values. Then we are going to find the genes that falls within those windows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing libraries\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we need to import the data files from \"C:\\Users\\aaisu\\Box\\Carter Lab\\Projects\\Anstep_GA_Yemen\\BA_Phased_Fst_12.14.2024\" into a dictionary of pandas data frames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully read file: EthiopianSampleNames.txtIndianSampleNames.txtYmEthInd_allChrs_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: EthiopianSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050201.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: EthiopianSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050202.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: EthiopianSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050203.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         145      0.161076  0.094346\n",
      "1  NC_050201.1      50001   150000         241      0.170661  0.113275\n",
      "2  NC_050201.1     100001   200000         314      0.276383  0.149935\n",
      "3  NC_050201.1     150001   250000         318      0.317534  0.149758\n",
      "4  NC_050201.1     200001   300000         364      0.161103  0.092637\n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         145      0.161076  0.094346\n",
      "1  NC_050201.1      50001   150000         241      0.170661  0.113275\n",
      "2  NC_050201.1     100001   200000         314      0.276383  0.149935\n",
      "3  NC_050201.1     150001   250000         318      0.317534  0.149758\n",
      "4  NC_050201.1     200001   300000         364      0.161103  0.092637\n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050202.1          1   100000         607      0.150210  0.094489\n",
      "1  NC_050202.1      50001   150000         912      0.144044  0.093122\n",
      "2  NC_050202.1     100001   200000         925      0.162536  0.104250\n",
      "3  NC_050202.1     150001   250000        1307      0.178556  0.108290\n",
      "4  NC_050202.1     200001   300000        2056      0.141020  0.090698\n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050203.1          1   100000         532      0.093384  0.062284\n",
      "1  NC_050203.1      50001   150000         854      0.136487  0.083668\n",
      "2  NC_050203.1     100001   200000        1504      0.222200  0.147422\n",
      "3  NC_050203.1     150001   250000        2265      0.213264  0.138717\n",
      "4  NC_050203.1     200001   300000        2630      0.190644  0.118143\n",
      "Successfully read file: EthiopianSampleNames.txtYemenSampleNames.txtYmEthInd_allChrs_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: EthiopianSampleNames.txtYemenSampleNames.txtYmEthInd_NC_050201.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: EthiopianSampleNames.txtYemenSampleNames.txtYmEthInd_NC_050202.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: EthiopianSampleNames.txtYemenSampleNames.txtYmEthInd_NC_050203.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         160      0.114982  0.063525\n",
      "1  NC_050201.1      50001   150000         276      0.114162  0.070737\n",
      "2  NC_050201.1     100001   200000         330      0.174455  0.097199\n",
      "3  NC_050201.1     150001   250000         324      0.191018  0.094940\n",
      "4  NC_050201.1     200001   300000         363      0.152308  0.079547\n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         160      0.114982  0.063525\n",
      "1  NC_050201.1      50001   150000         276      0.114162  0.070737\n",
      "2  NC_050201.1     100001   200000         330      0.174455  0.097199\n",
      "3  NC_050201.1     150001   250000         324      0.191018  0.094940\n",
      "4  NC_050201.1     200001   300000         363      0.152308  0.079547\n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050202.1          1   100000         447      0.118966  0.090075\n",
      "1  NC_050202.1      50001   150000         675      0.081213  0.069433\n",
      "2  NC_050202.1     100001   200000         760      0.110435  0.084474\n",
      "3  NC_050202.1     150001   250000        1115      0.152080  0.096992\n",
      "4  NC_050202.1     200001   300000        1848      0.121904  0.080411\n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050203.1          1   100000         630      0.033237  0.034087\n",
      "1  NC_050203.1      50001   150000         925      0.035785  0.034082\n",
      "2  NC_050203.1     100001   200000        1481      0.024166  0.028348\n",
      "3  NC_050203.1     150001   250000        2254      0.024409  0.025336\n",
      "4  NC_050203.1     200001   300000        2549      0.032190  0.028238\n",
      "Successfully read file: YemenSampleNames.txtIndianSampleNames.txtYmEthInd_allChrs_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: YemenSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050201.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: YemenSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050202.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: YemenSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050203.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         167      0.215941  0.052088\n",
      "1  NC_050201.1      50001   150000         263      0.252167  0.055406\n",
      "2  NC_050201.1     100001   200000         337      0.602737  0.149034\n",
      "3  NC_050201.1     150001   250000         369      0.636310  0.159827\n",
      "4  NC_050201.1     200001   300000         430      0.366817  0.106769\n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         167      0.215941  0.052088\n",
      "1  NC_050201.1      50001   150000         263      0.252167  0.055406\n",
      "2  NC_050201.1     100001   200000         337      0.602737  0.149034\n",
      "3  NC_050201.1     150001   250000         369      0.636310  0.159827\n",
      "4  NC_050201.1     200001   300000         430      0.366817  0.106769\n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050202.1          1   100000         580      0.178950  0.106299\n",
      "1  NC_050202.1      50001   150000         944      0.114165  0.066230\n",
      "2  NC_050202.1     100001   200000        1025      0.103182  0.061220\n",
      "3  NC_050202.1     150001   250000        1353      0.092715  0.057641\n",
      "4  NC_050202.1     200001   300000        2096      0.085762  0.057204\n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050203.1          1   100000         617      0.108284  0.071530\n",
      "1  NC_050203.1      50001   150000         959      0.107334  0.071892\n",
      "2  NC_050203.1     100001   200000        1582      0.203131  0.134041\n",
      "3  NC_050203.1     150001   250000        2266      0.192499  0.128995\n",
      "4  NC_050203.1     200001   300000        2621      0.162418  0.102490\n",
      "Successfully read file: YmEthSampleNames.txtIndianSampleNames.txtYmEthInd_allChrs_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: YmEthSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050201.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: YmEthSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050202.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "Successfully read file: YmEthSampleNames.txtIndianSampleNames.txtYmEthInd_NC_050203.1_BA_Phased.vcf.gz.vcf.gz_Fst.windowed.weir.fst\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         217      0.152742  0.053456\n",
      "1  NC_050201.1      50001   150000         351      0.154247  0.060080\n",
      "2  NC_050201.1     100001   200000         434      0.326410  0.104206\n",
      "3  NC_050201.1     150001   250000         453      0.383689  0.109380\n",
      "4  NC_050201.1     200001   300000         514      0.211837  0.072835\n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050201.1          1   100000         217      0.152742  0.053456\n",
      "1  NC_050201.1      50001   150000         351      0.154247  0.060080\n",
      "2  NC_050201.1     100001   200000         434      0.326410  0.104206\n",
      "3  NC_050201.1     150001   250000         453      0.383689  0.109380\n",
      "4  NC_050201.1     200001   300000         514      0.211837  0.072835\n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050202.1          1   100000         678      0.132929  0.085078\n",
      "1  NC_050202.1      50001   150000        1062      0.125855  0.078589\n",
      "2  NC_050202.1     100001   200000        1100      0.129402  0.079530\n",
      "3  NC_050202.1     150001   250000        1477      0.125711  0.077433\n",
      "4  NC_050202.1     200001   300000        2338      0.098943  0.065057\n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST\n",
      "0  NC_050203.1          1   100000         713      0.088768  0.045878\n",
      "1  NC_050203.1      50001   150000        1083      0.114712  0.059874\n",
      "2  NC_050203.1     100001   200000        1718      0.201649  0.120776\n",
      "3  NC_050203.1     150001   250000        2518      0.198458  0.118334\n",
      "4  NC_050203.1     200001   300000        2897      0.173545  0.100124\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def read_tab_delimited_files(directory_path, dict_name):\n",
    "    \"\"\"\n",
    "    Reads all tab-delimited text files in the specified directory into pandas DataFrames\n",
    "    and assigns them to the specified variable name in the global scope.\n",
    "\n",
    "    Args:\n",
    "        directory_path (str): Path to the directory containing the tab-delimited text files.\n",
    "        df_name (str): Name of the variable to store the resulting DataFrame dictionary.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    data_frames = {}\n",
    "\n",
    "    try:\n",
    "        # Iterate through files in the directory\n",
    "        for filename in os.listdir(directory_path):\n",
    "            # Check if the file has a .txt or .tsv extension (commonly used for tab-delimited files)\n",
    "            if filename.endswith(\".fst\"):\n",
    "                file_path = os.path.join(directory_path, filename)\n",
    "\n",
    "                # Read the file into a pandas DataFrame\n",
    "                df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "                # Defining data frame name\n",
    "                dfname = filename.split(\"_\")[1] + \"_\" + filename.split(\"_\")[2]\n",
    "\n",
    "                # Store the DataFrame in the dictionary\n",
    "                data_frames[dfname] = df\n",
    "\n",
    "                print(f\"Successfully read file: {filename}\")\n",
    "\n",
    "        # Assign the resulting dictionary to the given variable name in the global scope\n",
    "        globals()[dict_name] = data_frames\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "\n",
    "for folders in [\"Ethiopia_vs_India\", \"Ethiopia_vs_Yemen\", \"Yemen_vs_India\", \"YemenNEthiopia_vs_India\"]:\n",
    "    # Example usage\n",
    "    if __name__ == \"__main__\":\n",
    "        # Specify the directory containing your tab-delimited text files\n",
    "        directory = folders\n",
    "\n",
    "        # Specify the name of the variable to store the DataFrame dictionary\n",
    "        dictionary_name = folders\n",
    "\n",
    "        # Read the files and assign to the specified variable name\n",
    "        read_tab_delimited_files(directory, dictionary_name)\n",
    "\n",
    "        # Access the DataFrame dictionary dynamically\n",
    "        data_frames_dict = globals()[dictionary_name]\n",
    "\n",
    "        # Display the first few rows of each DataFrame\n",
    "        for file_name, df in data_frames_dict.items():\n",
    "            print(f\"\\nData from file: {file_name}\")\n",
    "            print(df.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding the genes lies within each Fst window\n",
    "\n",
    "In this step we are finding the genes that are overlapping with Fst Windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing the gene details\n",
    "\n",
    "pcg_filePath = \"Anstep_PCGs_1.12.2025.tsv\"\n",
    "\n",
    "PCG_det = pd.read_csv(pcg_filePath, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed Fst DataFrame: allChrs_BA\n",
      "Processed Fst DataFrame: NC_050201.1\n",
      "Processed Fst DataFrame: NC_050202.1\n",
      "Processed Fst DataFrame: NC_050203.1\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         145      0.161076  0.094346   \n",
      "1  NC_050201.1      50001   150000         241      0.170661  0.113275   \n",
      "2  NC_050201.1     100001   200000         314      0.276383  0.149935   \n",
      "3  NC_050201.1     150001   250000         318      0.317534  0.149758   \n",
      "4  NC_050201.1     200001   300000         364      0.161103  0.092637   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         145      0.161076  0.094346   \n",
      "1  NC_050201.1      50001   150000         241      0.170661  0.113275   \n",
      "2  NC_050201.1     100001   200000         314      0.276383  0.149935   \n",
      "3  NC_050201.1     150001   250000         318      0.317534  0.149758   \n",
      "4  NC_050201.1     200001   300000         364      0.161103  0.092637   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050202.1          1   100000         607      0.150210  0.094489   \n",
      "1  NC_050202.1      50001   150000         912      0.144044  0.093122   \n",
      "2  NC_050202.1     100001   200000         925      0.162536  0.104250   \n",
      "3  NC_050202.1     150001   250000        1307      0.178556  0.108290   \n",
      "4  NC_050202.1     200001   300000        2056      0.141020  0.090698   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118508044, LOC118508049, LOC118508043, LOC...  \n",
      "1  [LOC118508043, LOC118508048, LOC118508054, LOC...  \n",
      "2  [LOC118503348, LOC118503343, LOC118503351, LOC...  \n",
      "3  [LOC118503350, LOC118503352, LOC118503335, LOC...  \n",
      "4  [LOC118503365, LOC118503362, LOC118503359, LOC...  \n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050203.1          1   100000         532      0.093384  0.062284   \n",
      "1  NC_050203.1      50001   150000         854      0.136487  0.083668   \n",
      "2  NC_050203.1     100001   200000        1504      0.222200  0.147422   \n",
      "3  NC_050203.1     150001   250000        2265      0.213264  0.138717   \n",
      "4  NC_050203.1     200001   300000        2630      0.190644  0.118143   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118510147, LOC118510148, LOC118510149, LOC...  \n",
      "1  [LOC118508862, LOC118508863, LOC118508865, LOC...  \n",
      "2  [LOC118508865, LOC118508864, LOC118508866, LOC...  \n",
      "3                       [LOC118508871, LOC118508877]  \n",
      "4  [LOC118508879, LOC118508889, LOC118508888, LOC...  \n",
      "Processed Fst DataFrame: allChrs_BA\n",
      "Processed Fst DataFrame: NC_050201.1\n",
      "Processed Fst DataFrame: NC_050202.1\n",
      "Processed Fst DataFrame: NC_050203.1\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         160      0.114982  0.063525   \n",
      "1  NC_050201.1      50001   150000         276      0.114162  0.070737   \n",
      "2  NC_050201.1     100001   200000         330      0.174455  0.097199   \n",
      "3  NC_050201.1     150001   250000         324      0.191018  0.094940   \n",
      "4  NC_050201.1     200001   300000         363      0.152308  0.079547   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         160      0.114982  0.063525   \n",
      "1  NC_050201.1      50001   150000         276      0.114162  0.070737   \n",
      "2  NC_050201.1     100001   200000         330      0.174455  0.097199   \n",
      "3  NC_050201.1     150001   250000         324      0.191018  0.094940   \n",
      "4  NC_050201.1     200001   300000         363      0.152308  0.079547   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050202.1          1   100000         447      0.118966  0.090075   \n",
      "1  NC_050202.1      50001   150000         675      0.081213  0.069433   \n",
      "2  NC_050202.1     100001   200000         760      0.110435  0.084474   \n",
      "3  NC_050202.1     150001   250000        1115      0.152080  0.096992   \n",
      "4  NC_050202.1     200001   300000        1848      0.121904  0.080411   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118508044, LOC118508049, LOC118508043, LOC...  \n",
      "1  [LOC118508043, LOC118508048, LOC118508054, LOC...  \n",
      "2  [LOC118503348, LOC118503343, LOC118503351, LOC...  \n",
      "3  [LOC118503350, LOC118503352, LOC118503335, LOC...  \n",
      "4  [LOC118503365, LOC118503362, LOC118503359, LOC...  \n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050203.1          1   100000         630      0.033237  0.034087   \n",
      "1  NC_050203.1      50001   150000         925      0.035785  0.034082   \n",
      "2  NC_050203.1     100001   200000        1481      0.024166  0.028348   \n",
      "3  NC_050203.1     150001   250000        2254      0.024409  0.025336   \n",
      "4  NC_050203.1     200001   300000        2549      0.032190  0.028238   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118510147, LOC118510148, LOC118510149, LOC...  \n",
      "1  [LOC118508862, LOC118508863, LOC118508865, LOC...  \n",
      "2  [LOC118508865, LOC118508864, LOC118508866, LOC...  \n",
      "3                       [LOC118508871, LOC118508877]  \n",
      "4  [LOC118508879, LOC118508889, LOC118508888, LOC...  \n",
      "Processed Fst DataFrame: allChrs_BA\n",
      "Processed Fst DataFrame: NC_050201.1\n",
      "Processed Fst DataFrame: NC_050202.1\n",
      "Processed Fst DataFrame: NC_050203.1\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         167      0.215941  0.052088   \n",
      "1  NC_050201.1      50001   150000         263      0.252167  0.055406   \n",
      "2  NC_050201.1     100001   200000         337      0.602737  0.149034   \n",
      "3  NC_050201.1     150001   250000         369      0.636310  0.159827   \n",
      "4  NC_050201.1     200001   300000         430      0.366817  0.106769   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         167      0.215941  0.052088   \n",
      "1  NC_050201.1      50001   150000         263      0.252167  0.055406   \n",
      "2  NC_050201.1     100001   200000         337      0.602737  0.149034   \n",
      "3  NC_050201.1     150001   250000         369      0.636310  0.159827   \n",
      "4  NC_050201.1     200001   300000         430      0.366817  0.106769   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050202.1          1   100000         580      0.178950  0.106299   \n",
      "1  NC_050202.1      50001   150000         944      0.114165  0.066230   \n",
      "2  NC_050202.1     100001   200000        1025      0.103182  0.061220   \n",
      "3  NC_050202.1     150001   250000        1353      0.092715  0.057641   \n",
      "4  NC_050202.1     200001   300000        2096      0.085762  0.057204   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118508044, LOC118508049, LOC118508043, LOC...  \n",
      "1  [LOC118508043, LOC118508048, LOC118508054, LOC...  \n",
      "2  [LOC118503348, LOC118503343, LOC118503351, LOC...  \n",
      "3  [LOC118503350, LOC118503352, LOC118503335, LOC...  \n",
      "4  [LOC118503365, LOC118503362, LOC118503359, LOC...  \n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050203.1          1   100000         617      0.108284  0.071530   \n",
      "1  NC_050203.1      50001   150000         959      0.107334  0.071892   \n",
      "2  NC_050203.1     100001   200000        1582      0.203131  0.134041   \n",
      "3  NC_050203.1     150001   250000        2266      0.192499  0.128995   \n",
      "4  NC_050203.1     200001   300000        2621      0.162418  0.102490   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118510147, LOC118510148, LOC118510149, LOC...  \n",
      "1  [LOC118508862, LOC118508863, LOC118508865, LOC...  \n",
      "2  [LOC118508865, LOC118508864, LOC118508866, LOC...  \n",
      "3                       [LOC118508871, LOC118508877]  \n",
      "4  [LOC118508879, LOC118508889, LOC118508888, LOC...  \n",
      "Processed Fst DataFrame: allChrs_BA\n",
      "Processed Fst DataFrame: NC_050201.1\n",
      "Processed Fst DataFrame: NC_050202.1\n",
      "Processed Fst DataFrame: NC_050203.1\n",
      "\n",
      "Data from file: allChrs_BA\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         217      0.152742  0.053456   \n",
      "1  NC_050201.1      50001   150000         351      0.154247  0.060080   \n",
      "2  NC_050201.1     100001   200000         434      0.326410  0.104206   \n",
      "3  NC_050201.1     150001   250000         453      0.383689  0.109380   \n",
      "4  NC_050201.1     200001   300000         514      0.211837  0.072835   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050201.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050201.1          1   100000         217      0.152742  0.053456   \n",
      "1  NC_050201.1      50001   150000         351      0.154247  0.060080   \n",
      "2  NC_050201.1     100001   200000         434      0.326410  0.104206   \n",
      "3  NC_050201.1     150001   250000         453      0.383689  0.109380   \n",
      "4  NC_050201.1     200001   300000         514      0.211837  0.072835   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118511281, LOC118515097, LOC118513555, LOC...  \n",
      "1  [LOC118513625, LOC118513494, LOC118511019, LOC...  \n",
      "2  [LOC118513329, LOC118512064, LOC118512341, LOC...  \n",
      "3  [LOC118510517, LOC118512672, LOC118511740, LOC...  \n",
      "4  [LOC118511740, LOC118511794, LOC118515414, LOC...  \n",
      "\n",
      "Data from file: NC_050202.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050202.1          1   100000         678      0.132929  0.085078   \n",
      "1  NC_050202.1      50001   150000        1062      0.125855  0.078589   \n",
      "2  NC_050202.1     100001   200000        1100      0.129402  0.079530   \n",
      "3  NC_050202.1     150001   250000        1477      0.125711  0.077433   \n",
      "4  NC_050202.1     200001   300000        2338      0.098943  0.065057   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118508044, LOC118508049, LOC118508043, LOC...  \n",
      "1  [LOC118508043, LOC118508048, LOC118508054, LOC...  \n",
      "2  [LOC118503348, LOC118503343, LOC118503351, LOC...  \n",
      "3  [LOC118503350, LOC118503352, LOC118503335, LOC...  \n",
      "4  [LOC118503365, LOC118503362, LOC118503359, LOC...  \n",
      "\n",
      "Data from file: NC_050203.1\n",
      "         CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST  \\\n",
      "0  NC_050203.1          1   100000         713      0.088768  0.045878   \n",
      "1  NC_050203.1      50001   150000        1083      0.114712  0.059874   \n",
      "2  NC_050203.1     100001   200000        1718      0.201649  0.120776   \n",
      "3  NC_050203.1     150001   250000        2518      0.198458  0.118334   \n",
      "4  NC_050203.1     200001   300000        2897      0.173545  0.100124   \n",
      "\n",
      "                                               Genes  \n",
      "0  [LOC118510147, LOC118510148, LOC118510149, LOC...  \n",
      "1  [LOC118508862, LOC118508863, LOC118508865, LOC...  \n",
      "2  [LOC118508865, LOC118508864, LOC118508866, LOC...  \n",
      "3                       [LOC118508871, LOC118508877]  \n",
      "4  [LOC118508879, LOC118508889, LOC118508888, LOC...  \n"
     ]
    }
   ],
   "source": [
    "def map_genes_to_fst_windows(fst_dicts, gene_df):\n",
    "    \"\"\"\n",
    "    Maps gene symbols to Fst windows based on their positions and appends the list of gene names\n",
    "    to each Fst window in the Fst DataFrame.\n",
    "\n",
    "    Args:\n",
    "        fst_dicts (dict): Dictionary of DataFrames containing Fst windows.\n",
    "        gene_df (pd.DataFrame): DataFrame with gene details, including columns 'Begin', 'End', and 'Symbol'.\n",
    "\n",
    "    Returns:\n",
    "        dict: Updated dictionary of DataFrames with a new column 'Genes' containing lists of gene names.\n",
    "    \"\"\"\n",
    "    # Ensure 'Begin', 'End', and 'Symbol' columns exist in gene_df\n",
    "    required_columns = {'Begin', 'End', 'Symbol'}\n",
    "    if not required_columns.issubset(gene_df.columns):\n",
    "        raise ValueError(f\"The gene DataFrame must contain the columns: {required_columns}\")\n",
    "\n",
    "    for fst_name, fst_df in fst_dicts.items():\n",
    "        try:\n",
    "            # Create a new column for genes in the Fst DataFrame\n",
    "            fst_df['Genes'] = fst_df.apply(\n",
    "                lambda row: gene_df[\n",
    "                    (gene_df['Begin'] >= row['BIN_START']) & (gene_df['End'] <= row['BIN_END']) & (gene_df['Accession'] == row['CHROM'])\n",
    "                ]['Symbol'].tolist(), axis=1\n",
    "            )\n",
    "            print(f\"Processed Fst DataFrame: {fst_name}\")\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred while processing {fst_name}: {e}\")\n",
    "\n",
    "    return fst_dicts\n",
    "\n",
    "# Running the function on the data frames in dictionaries\n",
    "for dict in [\"Ethiopia_vs_India\", \"Ethiopia_vs_Yemen\", \"Yemen_vs_India\", \"YemenNEthiopia_vs_India\"]:\n",
    "    # Example usage\n",
    "    if __name__ == \"__main__\":\n",
    "\n",
    "        # Read the files and assign to the specified variable name\n",
    "        globals()[dict] = map_genes_to_fst_windows(globals()[dict], PCG_det)\n",
    "\n",
    "        # Access the DataFrame dictionary dynamically\n",
    "        data_frames_dict = globals()[dict]\n",
    "\n",
    "        # Display the first few rows of each DataFrame\n",
    "        for file_name, df in data_frames_dict.items():\n",
    "            print(f\"\\nData from file: {file_name}\")\n",
    "            print(df.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding the weighted Fst windows with significant Fst values\n",
    "\n",
    "In this step we are running a window down each data frame and adding a column to each data frame which specifies the category of the Fst value it belongs to based on the percentile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ethiopia_vs_India\n",
      "{75: np.float64(0.149153), 90: np.float64(0.19013300000000002), 95: np.float64(0.22529059999999998), 99: np.float64(0.32260711999999997)}\n",
      "Added 'percent_category' to DataFrame: allChrs_BA\n",
      "{75: np.float64(0.200837), 90: np.float64(0.26663040000000005), 95: np.float64(0.3193257999999998), 99: np.float64(0.5334296799999999)}\n",
      "Added 'percent_category' to DataFrame: NC_050201.1\n",
      "{75: np.float64(0.14612325), 90: np.float64(0.1887789), 95: np.float64(0.22163499999999997), 99: np.float64(0.31356327999999994)}\n",
      "Added 'percent_category' to DataFrame: NC_050202.1\n",
      "{75: np.float64(0.137368), 90: np.float64(0.16716940000000002), 95: np.float64(0.19068915), 99: np.float64(0.2530899099999999)}\n",
      "Added 'percent_category' to DataFrame: NC_050203.1\n",
      "Ethiopia_vs_Yemen\n",
      "{75: np.float64(0.08604115), 90: np.float64(0.11826420000000001), 95: np.float64(0.14395840000000001), 99: np.float64(0.21076234000000002)}\n",
      "Added 'percent_category' to DataFrame: allChrs_BA\n",
      "{75: np.float64(0.158885), 90: np.float64(0.20319700000000004), 95: np.float64(0.2364818999999999), 99: np.float64(0.3027480599999997)}\n",
      "Added 'percent_category' to DataFrame: NC_050201.1\n",
      "{75: np.float64(0.079228925), 90: np.float64(0.10512060000000001), 95: np.float64(0.12155094999999999), 99: np.float64(0.15234675999999997)}\n",
      "Added 'percent_category' to DataFrame: NC_050202.1\n",
      "{75: np.float64(0.077695725), 90: np.float64(0.10044370000000001), 95: np.float64(0.11744795), 99: np.float64(0.1496620499999999)}\n",
      "Added 'percent_category' to DataFrame: NC_050203.1\n",
      "Yemen_vs_India\n",
      "{75: np.float64(0.166711), 90: np.float64(0.2100852), 95: np.float64(0.25224889999999994), 99: np.float64(0.34760198000000003)}\n",
      "Added 'percent_category' to DataFrame: allChrs_BA\n",
      "{75: np.float64(0.250805), 90: np.float64(0.3140418), 95: np.float64(0.3736921999999996), 99: np.float64(0.6223013199999995)}\n",
      "Added 'percent_category' to DataFrame: NC_050201.1\n",
      "{75: np.float64(0.1644575), 90: np.float64(0.2013417), 95: np.float64(0.22758404999999995), 99: np.float64(0.30592309999999984)}\n",
      "Added 'percent_category' to DataFrame: NC_050202.1\n",
      "{75: np.float64(0.15189350000000001), 90: np.float64(0.1815398), 95: np.float64(0.206688), 99: np.float64(0.27386927)}\n",
      "Added 'percent_category' to DataFrame: NC_050203.1\n",
      "YemenNEthiopia_vs_India\n",
      "{75: np.float64(0.138241), 90: np.float64(0.176486), 95: np.float64(0.208071), 99: np.float64(0.31094466000000004)}\n",
      "Added 'percent_category' to DataFrame: allChrs_BA\n",
      "{75: np.float64(0.189373), 90: np.float64(0.24369480000000002), 95: np.float64(0.3052500999999999), 99: np.float64(0.50560946)}\n",
      "Added 'percent_category' to DataFrame: NC_050201.1\n",
      "{75: np.float64(0.13647274999999998), 90: np.float64(0.1764004), 95: np.float64(0.20810995), 99: np.float64(0.29758084999999995)}\n",
      "Added 'percent_category' to DataFrame: NC_050202.1\n",
      "{75: np.float64(0.1281965), 90: np.float64(0.1551237), 95: np.float64(0.17553069999999998), 99: np.float64(0.23898343999999994)}\n",
      "Added 'percent_category' to DataFrame: NC_050203.1\n"
     ]
    }
   ],
   "source": [
    "def add_percent_category(data_frames):\n",
    "    \"\"\"\n",
    "    Adds a column 'percent_category' to each DataFrame in the dictionary based on\n",
    "    the percentiles (75%, 90%, 95%, 99%) of the 'WEIGHTED_FST' column.\n",
    "\n",
    "    Args:\n",
    "        data_frames (dict): Dictionary where values are pandas DataFrames.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    for name, df in data_frames.items():\n",
    "        try:\n",
    "            # Calculate percentiles\n",
    "            percentiles = {\n",
    "                75: df['WEIGHTED_FST'].quantile(0.75),\n",
    "                90: df['WEIGHTED_FST'].quantile(0.90),\n",
    "                95: df['WEIGHTED_FST'].quantile(0.95),\n",
    "                99: df['WEIGHTED_FST'].quantile(0.99),\n",
    "            }\n",
    "\n",
    "            # Printing percentiles\n",
    "            print(percentiles)\n",
    "\n",
    "            # Define categorization function\n",
    "            def categorize(value):\n",
    "                if value >= percentiles[99]:\n",
    "                    return 'above_99_pct'\n",
    "                elif value >= percentiles[95]:\n",
    "                    return 'bw_95N99_pct'\n",
    "                elif value >= percentiles[90]:\n",
    "                    return 'bw_90N95_pct'\n",
    "                elif value >= percentiles[75]:\n",
    "                    return 'bw_75N90_pct'\n",
    "                else:\n",
    "                    return 'below_75_pct'\n",
    "\n",
    "            # Apply the categorization\n",
    "            df['percent_category'] = df['WEIGHTED_FST'].apply(categorize)\n",
    "            print(f\"Added 'percent_category' to DataFrame: {name}\")\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred while processing DataFrame {name}: {e}\")\n",
    "\n",
    "\n",
    "# Applying the function to assign the percentile category to each value in \n",
    "for dict in [\"Ethiopia_vs_India\", \"Ethiopia_vs_Yemen\", \"Yemen_vs_India\", \"YemenNEthiopia_vs_India\"]:\n",
    "\n",
    "    # Access the DataFrame dictionary dynamically\n",
    "    print(dict)\n",
    "    data_frames_dict = globals()[dict]\n",
    "    add_percent_category(data_frames_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dividing the dataframes based on the categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_data_frames_by_category(data_frames):\n",
    "    \"\"\"\n",
    "    Splits DataFrames in a dictionary into sub-dictionaries based on the 'percent_category' column.\n",
    "\n",
    "    Args:\n",
    "        data_frames (dict): Dictionary where values are pandas DataFrames.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary of dictionaries, where keys are DataFrame names and values\n",
    "              are dictionaries split by 'percent_category'.\n",
    "    \"\"\"\n",
    "    split_dict = {}\n",
    "\n",
    "    for name, df in data_frames.items():\n",
    "        try:\n",
    "            # Check if 'percent_category' column exists\n",
    "            if 'percent_category' not in df.columns:\n",
    "                print(f\"DataFrame {name} does not have 'percent_category' column. Skipping.\")\n",
    "                continue\n",
    "\n",
    "            # Create a sub-dictionary for this DataFrame\n",
    "            category_dict = {}\n",
    "\n",
    "            # Group by 'percent_category' and create a dictionary of DataFrames\n",
    "            for category, group in df.groupby('percent_category'):\n",
    "                category_dict[category] = group\n",
    "\n",
    "            # Add the sub-dictionary to the main dictionary\n",
    "            split_dict[name] = category_dict\n",
    "            print(f\"Split DataFrame {name} into categories.\")\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred while splitting DataFrame {name}: {e}\")\n",
    "\n",
    "    return split_dict\n",
    "\n",
    "\n",
    "# Running the function on the data frames in dictionaries\n",
    "for dict in [\"Ethiopia_vs_India\", \"Ethiopia_vs_Yemen\", \"Yemen_vs_India\", \"YemenNEthiopia_vs_India\"]:\n",
    "    # Example usage\n",
    "    if __name__ == \"__main__\":\n",
    "        # Specify the name of the variable to store the DataFrame dictionary\n",
    "        dictionary_name = dict + \"_splitOnPct\"\n",
    "\n",
    "        # Read the files and assign to the specified variable name\n",
    "        globals()[dictionary_name] = split_data_frames_by_category(globals()[dict])\n",
    "\n",
    "        # Access the DataFrame dictionary dynamically\n",
    "        data_frames_dict = globals()[dictionary_name]\n",
    "\n",
    "        # Display the result structure\n",
    "    for df_name, categories in data_frames_dict.items():\n",
    "        print(f\"\\nDataFrame: {df_name}\")\n",
    "        for category, df in categories.items():\n",
    "            print(f\"Category: {category}, Number of Rows: {len(df)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving the lists of genes for each category of percentiles for each chromosome of each population in separate text files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: Ethiopia_vs_India_splitOnPct_allChrs_BA_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_allChrs_BA_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_allChrs_BA_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_allChrs_BA_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_allChrs_BA_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050201.1_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050201.1_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050201.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050201.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050201.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050202.1_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050202.1_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050202.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050202.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050202.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050203.1_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050203.1_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050203.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050203.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_India_splitOnPct_NC_050203.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_allChrs_BA_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_allChrs_BA_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_allChrs_BA_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_allChrs_BA_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_allChrs_BA_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050201.1_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050201.1_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050201.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050201.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050201.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050202.1_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050202.1_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050202.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050202.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050202.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050203.1_above_99_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050203.1_below_75_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050203.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050203.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Ethiopia_vs_Yemen_splitOnPct_NC_050203.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_allChrs_BA_above_99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_allChrs_BA_below_75_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_allChrs_BA_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_allChrs_BA_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_allChrs_BA_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050201.1_above_99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050201.1_below_75_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050201.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050201.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050201.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050202.1_above_99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050202.1_below_75_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050202.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050202.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050202.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050203.1_above_99_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050203.1_below_75_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050203.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050203.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: Yemen_vs_India_splitOnPct_NC_050203.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_allChrs_BA_above_99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_allChrs_BA_below_75_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_allChrs_BA_bw_75N90_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_allChrs_BA_bw_90N95_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_allChrs_BA_bw_95N99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050201.1_above_99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050201.1_below_75_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050201.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050201.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050201.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050202.1_above_99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050202.1_below_75_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050202.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050202.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050202.1_bw_95N99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050203.1_above_99_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050203.1_below_75_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050203.1_bw_75N90_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050203.1_bw_90N95_pct_unique_genes.txt\n",
      "Saved: YemenNEthiopia_vs_India_splitOnPct_NC_050203.1_bw_95N99_pct_unique_genes.txt\n"
     ]
    }
   ],
   "source": [
    "# Define the folder where text files will be saved\n",
    "output_folder = \"Fst_unique_genes_text_files\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Running the function on the data frames in dictionaries\n",
    "for dict in [\"Ethiopia_vs_India_splitOnPct\", \"Ethiopia_vs_Yemen_splitOnPct\", \"Yemen_vs_India_splitOnPct\", \"YemenNEthiopia_vs_India_splitOnPct\"]:\n",
    "\n",
    "    # Access the DataFrame dictionary dynamically\n",
    "    data_frames_dict = globals()[dict]\n",
    "    \n",
    "    # Loop over the dictionary of dictionaries\n",
    "    for outer_key, inner_dict in data_frames_dict.items():\n",
    "        for inner_key, df in inner_dict.items():\n",
    "            # Extract all the unique genes from the 'Genes' column\n",
    "            unique_genes = set()\n",
    "            for gene_list in df['Genes']:\n",
    "                unique_genes.update(gene_list)  # Combine all genes into the set\n",
    "            \n",
    "            # Create a meaningful and sanitized file name\n",
    "            sanitized_outer_key = re.sub(r'[<>:\"/\\\\|?*]', '_', outer_key)\n",
    "            sanitized_inner_key = re.sub(r'[<>:\"/\\\\|?*]', '_', inner_key)\n",
    "            file_name = f\"{dict}_{sanitized_outer_key}_{sanitized_inner_key}_unique_genes.txt\".replace(\" \", \"_\")\n",
    "            file_path = os.path.join(output_folder, file_name)\n",
    "            \n",
    "            # Save the unique genes to the text file\n",
    "            with open(file_path, 'w') as f:\n",
    "                for gene in sorted(unique_genes):  # Sort genes alphabetically\n",
    "                    f.write(f\"{gene}\\n\")\n",
    "            \n",
    "            # Print a message to indicate progress\n",
    "            print(f\"Saved: {file_name}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "winbio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
