{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7df38026",
   "metadata": {},
   "source": [
    "# Converting VCF files to .Zarr files on Kodiak\n",
    "\n",
    "## Data preperation for genome-wide selection scans\n",
    "\n",
    "Installing necessary packages - scikit-allel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba1eb06-10ea-43a9-a2f5-2f288f221c98",
   "metadata": {},
   "source": [
    "Importing packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0e70af4b-91c1-491e-a919-f4247c1942bb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.3.8\n"
     ]
    }
   ],
   "source": [
    "import allel\n",
    "import sys\n",
    "print(allel.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4da0d4a7-cb36-46db-a603-fc85170cd43f",
   "metadata": {},
   "source": [
    "Reading VCF files (Use fields=* to read all the fields of the VCF file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "06d20436-debb-4bff-9962-2baa06f5ddb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "callset = allel.read_vcf('/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/YmEthInd_Combined/filtered_VCFs/YmEthInd_allChrs_fltpass.vcf', fileds = '*')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c35a6f6e-d71a-478a-bcca-a81e9858dc59",
   "metadata": {},
   "source": [
    "Checking the fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2e540e52-d2b8-4e40-928a-f89a60930dd4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['calldata/GT',\n",
       " 'samples',\n",
       " 'variants/ALT',\n",
       " 'variants/CHROM',\n",
       " 'variants/FILTER_PASS',\n",
       " 'variants/ID',\n",
       " 'variants/POS',\n",
       " 'variants/QUAL',\n",
       " 'variants/REF']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(callset.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ca4da59-3e88-497d-825e-4561c9e9e0e4",
   "metadata": {},
   "source": [
    "Getting the sample names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5d691a11-d387-4b58-8bf7-b978ad35f516",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SRR15257906', 'SRR15257907', 'SRR15257908', 'SRR15257909',\n",
       "       'SRR15257910', 'SRR15257911', 'SRR15257912', 'SRR15257913',\n",
       "       'SRR15257914', 'SRR15257915', 'SRR15293885', 'SRR15293886',\n",
       "       'SRR15293887', 'SRR15293888', 'SRR15293889', 'SRR15293890',\n",
       "       'SRR15293891', 'SRR15293892', 'SRR15293893', 'SRR15293894',\n",
       "       'X1296', 'X1307', 'X1402', 'X1403', 'X1404', 'X1408', 'X1409',\n",
       "       'X1410', 'X1415', 'X1416', 'X1417', 'X1419', 'X1420', 'X1421',\n",
       "       'X1423', 'X1424', 'X1425', 'X1580', 'X1581', 'X1583', 'X1585',\n",
       "       'X1586', 'X1587', 'X1604', 'X1605', 'X1673', 'X1676', 'X1679',\n",
       "       'X1680', 'X1735', 'X1736', 'X1738', 'X1740', 'X1742', 'X1743',\n",
       "       'X1747'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset['samples']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a884610-f974-4dbf-8477-b3032a712259",
   "metadata": {},
   "source": [
    "Getting genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "916decd5-a627-4a5a-accd-17620bb743b2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeArray shape=(12860511, 56, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">51</th><th style=\"text-align: center\">52</th><th style=\"text-align: center\">53</th><th style=\"text-align: center\">54</th><th style=\"text-align: center\">55</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">./.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"12\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12860508</th><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12860509</th><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12860510</th><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeArray shape=(12860511, 56, 2) dtype=int8>\n",
       "./. ./. ./. 1/1 ./. ... ./. 1/1 ./. ./. ./.\n",
       "./. ./. ./. ./. ./. ... ./. 0/0 ./. ./. ./.\n",
       "./. ./. ./. ./. 0/0 ... 0/0 ./. ./. 0/0 ./.\n",
       "...\n",
       "./. ./. 0/0 0/0 ./. ... ./. ./. ./. ./. ./.\n",
       "./. ./. 0/1 0/0 ./. ... ./. ./. ./. ./. ./.\n",
       "./. ./. 0/1 0/0 ./. ... ./. ./. ./. ./. ./."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt = allel.GenotypeArray(callset['calldata/GT'])\n",
    "gt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "37a43c4c-888c-4548-ae58-9d0214f764bf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([3, 2, 0, ..., 0, 1, 1])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt.is_het()\n",
    "gt.count_het(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4cbf41a2-bfac-491c-a2a7-fd897835cdc0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;AlleleCountsArray shape=(12860511, 7) dtype=int32&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">5</th><th style=\"text-align: center\">6</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">13</td><td style=\"text-align: center\">21</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">42</td><td style=\"text-align: center\"> 2</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">46</td><td style=\"text-align: center\"> 2</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"8\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12860508</th><td style=\"text-align: center\">12</td><td style=\"text-align: center\"> 2</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12860509</th><td style=\"text-align: center\"> 9</td><td style=\"text-align: center\"> 3</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12860510</th><td style=\"text-align: center\"> 9</td><td style=\"text-align: center\"> 3</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td><td style=\"text-align: center\"> 0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<AlleleCountsArray shape=(12860511, 7) dtype=int32>\n",
       "13 21  0  0  0  0  0\n",
       "42  2  0  0  0  0  0\n",
       "46  2  0  0  0  0  0\n",
       "...\n",
       "12  2  0  0  0  0  0\n",
       " 9  3  0  0  0  0  0\n",
       " 9  3  0  0  0  0  0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ac = gt.count_alleles()\n",
    "ac"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebfbace9-4936-46d7-8d31-a3cf23eda83a",
   "metadata": {},
   "source": [
    "## Reading VCF files to Zarr files\n",
    "Defining paths for input and output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd343ec0-3ad1-42bb-9f4e-58a1570f1a94",
   "metadata": {},
   "source": [
    "Installing Zarr Package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0d8db5db-954f-40fe-bcc6-5b306a74316d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: zarr in /home/gunarathnai/.local/lib/python3.9/site-packages (2.18.2)\n",
      "Requirement already satisfied: numpy>=1.23 in /home/gunarathnai/.local/lib/python3.9/site-packages (from zarr) (1.26.4)\n",
      "Requirement already satisfied: asciitree in /home/gunarathnai/.local/lib/python3.9/site-packages (from zarr) (0.3.3)\n",
      "Requirement already satisfied: numcodecs>=0.10.0 in /home/gunarathnai/.local/lib/python3.9/site-packages (from zarr) (0.12.1)\n",
      "Requirement already satisfied: fasteners in /home/gunarathnai/.local/lib/python3.9/site-packages (from zarr) (0.19)\n",
      "\u001b[33mWARNING: You are using pip version 21.3.1; however, version 24.3.1 is available.\n",
      "You should consider upgrading via the '/cm/local/apps/python39/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: numcodecs in /home/gunarathnai/.local/lib/python3.9/site-packages (0.12.1)\n",
      "Requirement already satisfied: numpy>=1.7 in /home/gunarathnai/.local/lib/python3.9/site-packages (from numcodecs) (1.26.4)\n",
      "\u001b[33mWARNING: You are using pip version 21.3.1; however, version 24.3.1 is available.\n",
      "You should consider upgrading via the '/cm/local/apps/python39/bin/python3 -m pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install zarr\n",
    "!pip install numcodecs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39dbd8aa-97db-4419-9c4b-16d6d4073ca0",
   "metadata": {},
   "source": [
    "Importing Zarr library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "24091faf-0297-4b69-9547-80254e808199",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import zarr\n",
    "import numcodecs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "56090c55-b46b-4a17-b41f-fa83241f0112",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "vcf_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/YmEthInd_Combined/filtered_VCFs/YmEthInd_allChrs_fltpass.vcf'\n",
    "zarr_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/Zarr_files/YmEthInd_allChrs_fltpass.zarr'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77b71a34-7947-4500-a097-f8f235afcbe8",
   "metadata": {},
   "source": [
    "Running VCF to Zarr conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "96f11375-aae8-485a-8d95-b60e27940274",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "allel.vcf_to_zarr(vcf_path, zarr_path, fields='*', overwrite =True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f19add9d-820a-46de-b478-489e410dba0a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<zarr.hierarchy.Group '/' read-only>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset_zarr = zarr.open_group('YmEthInd_allChrs_fltpass.zarr', mode='r')\n",
    "callset_zarr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfb7be11-6d3d-48cf-804d-b076f0791629",
   "metadata": {},
   "source": [
    "Exploring the hierarchy of groups and arrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6cdfac39-bce8-473c-98c0-acc5c1c17ee6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "No module named 'ipytree': Run `pip install zarr[jupyter]` or `conda install ipytree`to get the required ipytree dependency for displaying the tree widget. If using jupyterlab<3, you also need to run `jupyter labextension install ipytree`",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/zarr/util.py:494\u001b[0m, in \u001b[0;36mtree_widget\u001b[0;34m(group, expand, level)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 494\u001b[0m     \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mipytree\u001b[39;00m\n\u001b[1;32m    495\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'ipytree'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "File \u001b[0;32m/cm/shared/apps/jupyter/current/lib/python3.9/site-packages/IPython/core/formatters.py:974\u001b[0m, in \u001b[0;36mMimeBundleFormatter.__call__\u001b[0;34m(self, obj, include, exclude)\u001b[0m\n\u001b[1;32m    971\u001b[0m     method \u001b[38;5;241m=\u001b[39m get_real_method(obj, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprint_method)\n\u001b[1;32m    973\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m method \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 974\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[43minclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minclude\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexclude\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    975\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    976\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/zarr/util.py:553\u001b[0m, in \u001b[0;36mTreeViewer._repr_mimebundle_\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    552\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_repr_mimebundle_\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m--> 553\u001b[0m     tree \u001b[38;5;241m=\u001b[39m \u001b[43mtree_widget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgroup\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexpand\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexpand\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    554\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m tree\u001b[38;5;241m.\u001b[39m_repr_mimebundle_(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/zarr/util.py:496\u001b[0m, in \u001b[0;36mtree_widget\u001b[0;34m(group, expand, level)\u001b[0m\n\u001b[1;32m    494\u001b[0m     \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mipytree\u001b[39;00m\n\u001b[1;32m    495\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 496\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(\n\u001b[1;32m    497\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m: Run `pip install zarr[jupyter]` or `conda install ipytree`\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    498\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mto get the required ipytree dependency for displaying the tree \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    499\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwidget. If using jupyterlab<3, you also need to run \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    500\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m`jupyter labextension install ipytree`\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    501\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n\u001b[1;32m    503\u001b[0m result \u001b[38;5;241m=\u001b[39m ipytree\u001b[38;5;241m.\u001b[39mTree()\n\u001b[1;32m    504\u001b[0m root \u001b[38;5;241m=\u001b[39m TreeNode(group, level\u001b[38;5;241m=\u001b[39mlevel)\n",
      "\u001b[0;31mImportError\u001b[0m: No module named 'ipytree': Run `pip install zarr[jupyter]` or `conda install ipytree`to get the required ipytree dependency for displaying the tree widget. If using jupyterlab<3, you also need to run `jupyter labextension install ipytree`"
     ]
    },
    {
     "data": {
      "text/plain": [
       "/\n",
       " ├── calldata\n",
       " │   ├── AD (12860511, 56, 4) int16\n",
       " │   ├── DP (12860511, 56) int16\n",
       " │   ├── GQ (12860511, 56) int8\n",
       " │   ├── GT (12860511, 56, 2) int8\n",
       " │   ├── MIN_DP (12860511, 56) int32\n",
       " │   ├── PGT (12860511, 56) object\n",
       " │   ├── PID (12860511, 56) object\n",
       " │   ├── PL (12860511, 56, 3) int32\n",
       " │   ├── PS (12860511, 56) int32\n",
       " │   ├── RGQ (12860511, 56) int32\n",
       " │   └── SB (12860511, 56, 4) int32\n",
       " ├── samples (56,) object\n",
       " └── variants\n",
       "     ├── AC (12860511, 3) int32\n",
       "     ├── AF (12860511, 3) float32\n",
       "     ├── ALT (12860511, 3) object\n",
       "     ├── AN (12860511,) int32\n",
       "     ├── AS_BaseQRankSum (12860511, 3) float32\n",
       "     ├── AS_FS (12860511, 3) float32\n",
       "     ├── AS_InbreedingCoeff (12860511, 3) float32\n",
       "     ├── AS_MQ (12860511, 3) float32\n",
       "     ├── AS_MQRankSum (12860511, 3) float32\n",
       "     ├── AS_QD (12860511, 3) float32\n",
       "     ├── AS_ReadPosRankSum (12860511, 3) float32\n",
       "     ├── AS_SOR (12860511, 3) float32\n",
       "     ├── BaseQRankSum (12860511,) float32\n",
       "     ├── CHROM (12860511,) object\n",
       "     ├── DP (12860511,) int32\n",
       "     ├── END (12860511,) int32\n",
       "     ├── ExcessHet (12860511,) float32\n",
       "     ├── FILTER_LowDepth (12860511,) bool\n",
       "     ├── FILTER_LowQD (12860511,) bool\n",
       "     ├── FILTER_LowQual (12860511,) bool\n",
       "     ├── FILTER_MappingQuality (12860511,) bool\n",
       "     ├── FILTER_PASS (12860511,) bool\n",
       "     ├── FILTER_ReadPosRank (12860511,) bool\n",
       "     ├── FILTER_SOR (12860511,) bool\n",
       "     ├── FILTER_StrandBias (12860511,) bool\n",
       "     ├── FS (12860511,) float32\n",
       "     ├── ID (12860511,) object\n",
       "     ├── InbreedingCoeff (12860511,) float32\n",
       "     ├── MLEAC (12860511, 3) int32\n",
       "     ├── MLEAF (12860511, 3) float32\n",
       "     ├── MQ (12860511,) float32\n",
       "     ├── MQRankSum (12860511,) float32\n",
       "     ├── POS (12860511,) int32\n",
       "     ├── QD (12860511,) float32\n",
       "     ├── QUAL (12860511,) float32\n",
       "     ├── RAW_MQandDP (12860511, 2) int32\n",
       "     ├── REF (12860511,) object\n",
       "     ├── ReadPosRankSum (12860511,) float32\n",
       "     ├── SOR (12860511,) float32\n",
       "     ├── altlen (12860511, 3) int32\n",
       "     ├── is_snp (12860511,) bool\n",
       "     └── numalt (12860511,) int32"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset_zarr.tree(expand=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e696088b-3332-42e1-9fbe-3d166906ecbb",
   "metadata": {},
   "source": [
    "# Importing phased biallelic variants for each chomosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5bd35faf-480a-4aef-9dec-be1fc82adde2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Chromosome X\n",
    "chrx_vcf_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/YmEthInd_Combined/biallelic_phased_VCFs/YmEthInd_NC_050201.1_BA_Phased.vcf.gz.vcf.gz'\n",
    "chrx_zarr_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/Zarr_files/YmEthInd_Chrx_BAP.zarr'\n",
    "allel.vcf_to_zarr(chrx_vcf_path, chrx_zarr_path, fields='*', overwrite =True)\n",
    "\n",
    "# Chromosome 2\n",
    "chr2_vcf_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/YmEthInd_Combined/biallelic_phased_VCFs/YmEthInd_NC_050202.1_BA_Phased.vcf.gz.vcf.gz'\n",
    "chr2_zarr_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/Zarr_files/YmEthInd_Chr2_BAP.zarr'\n",
    "allel.vcf_to_zarr(chr2_vcf_path, chr2_zarr_path, fields='*', overwrite =True)\n",
    "\n",
    "# Chromosome 3\n",
    "chr3_vcf_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/YmEthInd_Combined/biallelic_phased_VCFs/YmEthInd_NC_050203.1_BA_Phased.vcf.gz.vcf.gz'\n",
    "chr3_zarr_path = '/data/gunarathnai/Ans_GA/gatk_variants/YmEthSmlInd_variants/Zarr_files/YmEthInd_Chr3_BAP.zarr'\n",
    "allel.vcf_to_zarr(chr3_vcf_path, chr3_zarr_path, fields='*', overwrite =True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6544f731-5aaf-4d5b-8167-019e4a802f17",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
